### ISSUE TITLE
Throw a better error message for out of range index 

### ISSUE THREAD

**Issue by zohimchandani** [NVIDIA/cuda-quantum#683]
### Required prerequisites

- [X] Make sure you've read the [documentation](https://nvidia.github.io/cuda-quantum/latest). Your issue may be addressed there.
- [X] Search the [issue tracker](https://github.com/NVIDIA/cuda-quantum/issues) to verify that this hasn't already been reported. +1 or comment there if it has.
- [X] If possible, make a PR with a failing test to give us a starting point to work on!

### Describe the bug

When addressing qubits out of the range of the ones available, the error message shown is: 

```python3: /workspaces/cuda-quantum/runtime/nvqir/QIRTypes.cpp:22: int8_t* Array::operator[](std::size_t): Assertion `static_cast<uint64_t>(index * element_size_bytes) < storage.size() && "Provided index for Array::operator[] not valid."' failed.
Aborted (core dumped)```

Can we improve this please? Thanks.  

### Steps to reproduce the bug

NA

### Expected behavior

NA

### Is this a regression? If it is, put the last known working version (or commit) here.

Not a regression

### Environment

- **CUDA Quantum version**: 12
- **Python version**: 3.10
- **C++ compiler**: -
- **Operating system**: Linux

**Comment by anthony-santana** [comment]
Example to reproduce error
```
import cudaq 

kernel = cudaq.make_kernel()
qubits = kernel.qalloc(5)

kernel.x(qubits[10])

print(kernel)

cudaq.sample(kernel)

```

**Comment by schweitzpgi** [comment]
Reproducer? Thanks.

**Comment by schweitzpgi** [comment]
> Example to reproduce error
> 
> ```
> import cudaq 
> 
> kernel = cudaq.make_kernel()
> qubits = kernel.qalloc(5)
> 
> kernel.x(qubits[10])
> 
> print(kernel)
> 
> cudaq.sample(kernel)
> ```

Does the python side do any sort of checking?

**Comment by anthony-santana** [comment]
> > Example to reproduce error
> > ```
> > import cudaq 
> > 
> > kernel = cudaq.make_kernel()
> > qubits = kernel.qalloc(5)
> > 
> > kernel.x(qubits[10])
> > 
> > print(kernel)
> > 
> > cudaq.sample(kernel)
> > ```
> 
> Does the python side do any sort of checking?

No, just filed an issue to throw an error during compilation as well #684

**Comment by bmhowe23** [comment]
I took a stab at resolving this in PR #688. Let me know what you think.

**Comment by bmhowe23** [comment]
After merging #688, the example in this issue produces the following, so I am going to close this issue.
```
loc("<builder>":1:1): error: 'quake.extract_ref' op invalid index [10] because >= size [5]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: cudaq::builder failed to JIT compile the Quake representation.
```

Resolved by #688.
